apps/heft/src/plugins/TypeScriptPlugin/Eslint.ts

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import * as crypto from 'crypto';
+import * as TEslint from 'eslint';
+
+import { LinterBase, ILinterBaseOptions } from './LinterBase';
+import { IExtendedSourceFile } from './internalTypings/TypeScriptInternals';
+import { IColorableSequence, Colors } from '@rushstack/node-core-library';
+
+interface IEslintOptions extends ILinterBaseOptions {
+  eslintPackagePath: string;
+}
+
+export class Eslint extends LinterBase<TEslint.ESLint.LintResult> {
+  private readonly _eslintPackage: typeof TEslint;
+
+  private _eslintCli: TEslint.CLIEngine;
+  private _eslint: TEslint.ESLint;
+  private _eslintBaseConfiguration: any; // eslint-disable-line @typescript-eslint/no-explicit-any
+  private _lintResult: TEslint.ESLint.LintResult[];
+
+  public constructor(options: IEslintOptions) {
+    super('eslint', options);
+
+    this._eslintPackage = require(options.eslintPackagePath);
+  }
+
+  public printVersionHeader(): void {
+    this._terminal.writeLine(`Using ESLint version ${this._eslintPackage.Linter.version}`);
+  }
+
+  public reportFailures(): void {
+    const eslintFailureLogMessages: (string | IColorableSequence)[][] = [];
+
+    for (const eslintFileResult of this._lintResult) {
+      const buildFolderRelativeFilePath: string = path.relative(
+        this._buildFolderPath,
+        eslintFileResult.filePath
+      );
+      for (const message of eslintFileResult.messages) {
+        // https://eslint.org/docs/developer-guide/nodejs-api#◆-lintmessage-type
+        const severity: string = message.severity === 1 ? 'WARNING' : 'ERROR';
```

Comment:
![Capture](https://user-images.githubusercontent.com/4673363/86744305-607abe80-bfee-11ea-885c-684486d0b7ac.png)

There are some formatting issues with the above output:
- The `@rushstack/eslint-config` package classifies every lint issue as an "error". We need to fix this, since nearly all lint issues are not serious and should be warnings. It's an easy fix that you could possibly do in this PR.
- Here Heft is remapping every lint issue to be a warning regardless of its severity, presumably to compensate for that.  If we keep this approach, then we cannot print `ERROR`.
- API Extractor's warnings are printed in red

The **important UI design principles** are:
- Log messages that call themselves "error" must be treated as errors (print to STDERR and return nonzero a exit code so Rush fails)
- Log messages that call themselves "warning" must be treated as warnings (print to STDERR and not change the exit code)
- The red color should never be used for anything that is not an actual error
- The yellow color should never be used for anything that is not an actual warning

Reply:
If we want to summarize the error/warning counts, they really should come at the very end, something like:
```
 ---- Bundle finished (800ms) ----
-------------------- Finished (2.55s) --------------------
Project: heft-test-01@1.0.0
ESLint reported 3 warnings <--- yellow
TSLint reported 3 errors and 4 warnings <--- red
API Extractor reported 2 errors <--- red
```
If a task (e.g. TypeScript) completed successfully, I wouldn't include it in the summary at all, because that is just taking up space without telling us anything we don't already know.  However, on 100% success we could show something like this:
```
-------------------- Finished (2.55s) --------------------
Project: heft-test-01@1.0.0
All tasks completed without warnings. <--- green
```

It is weird for "Finished" to be colored green when there are warnings. Maybe don't color it at all.

Reply:
On a separate topic, I'm still strongly in favor of [stream-collator](https://www.npmjs.com/package/@rushstack/stream-collator) which will eliminate all these `[api-extractor] ` prefixes cluttering up the output. Our UI design would instead show coherent chunks of reporting for individual tasks, sort of like you're doing with `Encountered 2 ESLint errors:`

Reply:
stream-collator doesn't handle watch processes well. Perhaps we need two modes? An important part of a watch mode is seeing *something* getting logged, telling you that the process hasn't frozen.

I'm not completely sold on separating blocks of logs into contiguous streams as a good UI. In Rush it makes sense because it's separate projects that are effectively independent, so their parallel builds are irrelevant from a logging perspective. The same doesn't apply to a truly parallelized build of a single project.

I'm going to rework the error reporting in a separate PR to follow some of the ideas you've posted here, and then I think we should do subsequent PRs to rework the rest of the logging UI.